File: src/algorithms/python/Path_Sum.py
# -*- coding: utf-8 -*-


# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    def dfs(self, root, cnt):
        if root.left:
            self.dfs(root.left, cnt + root.val)
        if root.right:
            self.dfs(root.right, cnt + root.val)
        if root.left is None and root.right is None:
            if cnt + root.val == self.sum:
                self.flag = True

    # @param {TreeNode} root
    # @param {integer} sum
    # @return {boolean}
    def hasPathSum(self, root, sum):
        if root is None:
            return False
        self.flag = False
        self.sum = sum
        self.dfs(root, 0)
        return self.flag


if __name__ == "__main__":
    solution = Solution()
    print solution.hasPathSum(None, 0)
    node = TreeNode(1)
    print solution.hasPathSum(node, 1)

File: src/algorithms/cpp/Path_Sum.cpp
#include <iostream>
#include <cstdio>
#include <cstring>
#include <cmath>
#include <cstdlib>
#include <algorithm>
#include <vector>
#include <map>
#include <set>
using namespace std;
#define out(v) cerr << #v << ": " << (v) << endl
#define SZ(v) ((int)(v).size())
const int maxint = -1u>>1;
template <class T> bool get_max(T& a, const T &b) {return b > a? a = b, 1: 0;}
template <class T> bool get_min(T& a, const T &b) {return b < a? a = b, 1: 0;}

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
private:
    bool check(TreeNode *root, int sum) {
        bool flag = false;
        if (root->left)
            flag = flag || check(root->left, sum - root->val);
        if (root->right)
            flag = flag || check(root->right, sum - root->val);
        if (root->left == NULL && root->right == NULL && sum - root->val == 0)
            return true;
        return flag;
    }

public:
    bool hasPathSum(TreeNode *root, int sum) {
        if (root == NULL)   return false;
        else                return check(root, sum);
    }
};

int main() {
    return 0;
}
